src/common/transformations/src/transformations/resolve_names_collisions.cpp

```diff
@@ -21,62 +21,39 @@ bool ov::pass::ResolveNameCollisions::run_on_model(const std::shared_ptr<ov::Mod
     std::unordered_map<std::string, std::list<Node*>> visited_nodes;
 
     for (const auto& node : model->get_ordered_ops()) {
-        // Detect names collisions only for nodes with autogenerated names
+        // Collect a names collision map for all nodes in the graph
         const auto& friendly_name = node->get_friendly_name();
         visited_nodes[friendly_name].emplace_back(node.get());
     }
 
-    for (const auto& l_nodes : visited_nodes) {
-        if (l_nodes.second.size() == 1)
-            continue;
-        const size_t nodes_size = l_nodes.second.size();
-        bool has_public_node = false;  // Parameter, Result ans Sinks
-        size_t i(0);
-        for (auto* node : l_nodes.second) {
-            i++;
-            // Skip the last node if we don't have public nodes with collisions
-            if (i == nodes_size && !has_public_node)
-                break;
-            if (dynamic_cast<const ov::op::v0::Result*>(node)) {
-                // Result is a service node
-                continue;
-            }
-            if (dynamic_cast<const ov::op::Sink*>(node) || dynamic_cast<const ov::op::v0::Parameter*>(node)) {
-                // Resolve names for public ops with autogenerated name
-                if (node->m_friendly_name.empty())
-                    nodes_with_conflicts.emplace_back(node);
-                has_public_node = true;
-                continue;
-            } else {
-                // For result we need to avoid changes in previous operation
-                bool is_public = false;
-                for (const auto& output : node->outputs()) {
-                    for (const auto input : output.get_target_inputs()) {
-                        if (dynamic_cast<const ov::op::v0::Result*>(input.get_node())) {
-                            has_public_node = true;
-                            is_public = true;
+    for (const auto& it : visited_nodes) {
+        const auto& same_named_ops = it.second;
+        if (same_named_ops.size() > 1) {
+            int64_t cnt = 2;
+
+            for (const auto& op : same_named_ops) {
+                // check if op has OV autogenerated friendly name. unique and friendly names have to be equal.
+                bool is_autogenerated = op->get_friendly_name() == op->get_name();
+                if (is_autogenerated) {
+                    // add a prefix "_counter" to the autogenerated name to make it unique.
+                    auto new_name = op->get_friendly_name() + "_" + std::to_string(cnt++);
+
+                    int64_t safety_stop = 10000;
+                    bool safety_stop_happened = false;
+                    while (visited_nodes.find(new_name) != visited_nodes.end()) {
+                        new_name = op->get_friendly_name() + "_" + std::to_string(cnt++);
+                        if (cnt > safety_stop) {
+                            safety_stop_happened = true;
```

Comment:
Can we do `op->set_friendly_name(new_name)` inside this condition?

Reply:
I don't think so, safety_stop stop is just to avoid infinity loop
theoretically it should never happen